torchtune/utils/_profiler.py

```diff
@@ -4,47 +4,358 @@
+    if not os.path.exists(curr_trace_dir):
+        os.makedirs(curr_trace_dir, exist_ok=True)
+
+    # Export chrome / tensorboard trace
+    if rank == 0:
+        log.info(f"Dumping traces at step {prof.step_num}")
+    begin = time.monotonic()
+
+    # Use tensorboard trace handler rather than directly exporting chrome traces since
+    # tensorboard doesn't seem to be able to parse traces with prof.export_chrome_trace
+    exporter = tensorboard_trace_handler(
+        curr_trace_dir, worker_name=f"rank{rank}", use_gzip=True
+    )
+    exporter(prof)
+
+    if rank == 0:
+        log.info(f"Finished dumping traces in {time.monotonic() - begin:.2f} seconds")
+
+    # Memory timeline sometimes fails to export
+    if prof.profile_memory:
+        if rank == 0:
+            try:
+                prof.export_memory_timeline(
+                    f"{curr_trace_dir}/rank{rank}_memory-timeline.html"
+                )
+            except Exception as e:
+                log.warn(f" Failed to export memory timeline: {e}")
+
+    # Dump stack traces
+    if prof.with_stack:
+        prof.export_stacks(f"{curr_trace_dir}/rank{rank}_stacks.txt", metric=metric)
+
+    # Export event averages
+    key_avgs = prof.key_averages(
+        group_by_input_shape=prof.record_shapes, group_by_stack_n=5
+    ).table(sort_by=metric, row_limit=row_limit)
+    with open(f"{curr_trace_dir}/rank{rank}_key_averages.txt", "w") as f:
+        print(key_avgs, file=f)
+    if rank == 0:
+        log.info(f"Saving profiling results to {curr_trace_dir}")
+
+    # TODO: Is this necessary?
+    # see https://github.com/pytorch/torchtitan/blob/3050098dcee4901d88c712f9e8e9703d1735a29b/torchtitan/profiling.py#L48
+    if world_size > 1:
+        torch.distributed.barrier()
+
+
+class FakeProfiler:
+    """
+    Drop-in replacement for torch.profiler.profile that functions as a nullcontext / object
+    with no-op methods for ``start``, ``stop``, and ``step``.
+
+    This is helpful for instrumenting profiling in a recipe without requiring changes to the
+    code independent of whether profiling is on / off.
+
+    E.g.,
+
+    .. code-block:: python
+
+        profiler = FakeProfiler()
+        #profiler = torch.profiler.profile()
+
+        # Below is same regardless of profiler object type
+        with profiler as prof:
+            for epoch in epochs:
+                for batch in batches:
+                    train.step()
+                    prof.step()
+
+    """
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        pass
+
+    def start(self):
+        pass
+
+    def stop(self):
+        pass
+
+    def step(self):
+        pass
+
+
+def setup_torch_profiler(
+    enabled: bool = False,
+    cpu: bool = True,
+    cuda: bool = True,
+    profile_memory: bool = False,
+    with_stack: bool = False,
+    record_shapes: bool = False,
+    with_flops: bool = False,
+    wait: Optional[int] = None,
```

Comment:
If we altered this to support `config.initialize(setup_torch_profiler)` from the recipe, then all of the schedule args could be renamed to "schedule_warmup", "schedule_active", etc allowing for a flat list of profiler configs. All of these config options that are None would be optional by default in the config and if they get a None value, they can be set to the default value in the init.

Reply:
What you're describing is more in-line with my original PR, where most of the code in `_setup_profiler` was abstracted away from the user in `setup_torch_profiler`.  

Per the comments from @RdoubleA and @ebsmothers, I refactored the original `setup_torch_profiler` such that most of the setup was extracted from `setup_torch_profiler` to the `recipe`, which admittedly results in more code duplication and config parsing on the part of the recipe writer / user. 

Not sure what direction to take at this point.

Reply:
I am sorry for the confusion. In the original version, there was a lot of config parsing happening inside of the util, which we don't want for a utility since our principle is we want each function to work as code and to not assume that a config will be involved. So I agree with all the changes to to remove the config parsing from the util. But I believe that we can still get that same simplicity that you originally proposed by defining all the arguments inside of the __init__ instead of taking in a Dict. Something like this

```python
def setup_torch_profiler(
        self,
        *,
        enabled: bool = True
        output_dir: str = "profiler_output",
        #`torch.profiler.ProfilerActivity` types to trace
        activity_types: List[str] = ACTIVITY_TYPES.keys(),
        #trace options passed to `torch.profiler.profile`
        profile_memory: bool =  False
  	with_stack: bool =  False
  	record_shapes: bool = True
  	with_flops: bool =  True
  	# `torch.profiler.schedule` options
	schedule_wait: int = 5
	schedule_warmup: int = 5
        schedule_active: int = 2
	schedule_repeat: int = 1
        log_cfg: bool = False
) -> ContextManager:
```

Then `config.initialize(setup_torch_profiler)` could initialize it from the config automatically and setup_torch_profiler works as a stand alone function. Does this make sense? I'll tag @RdoubleA and @ebsmothers here to make sure I'm not contradicting them, because I don't want you to go back and forth on work.

Reply:
Yeah sorry for the thrash here @jeromeku. I think @pbontrager's last comment here generally makes sense to me. Basically we can define reasonable defaults for all params inside of `setup_torch_profiler`. That way we don't have to go through the work of parsing everything and setting defaults inside the recipe. Then the config would look like

```
profiler:
  _component_: torchtune.utils._profiler.setup_torch_profiler
  enabled: True
  cpu: True
  cuda: True
  ...
```

and in the recipe you can just call `config.instantiate(cfg.profiler)` in the recipe's `_setup_profiler` method. You then only have to check that the profiler field is defined, but by putting defaults in the `setup_torch_profiler` utility you don't really have to do a bunch of sanity checks on missing fields in the config.

Reply:
Yes to everything Evan said with the small change, that we wouldn't need a _setup_profiler method then and could just put it directly in the setup method.